File: TuCancha_Microservicios/Reservas/src/main/java/com/example/Reservas/model/Reserva.java
package com.example.Reservas.model;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "reservas")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Schema(description = "Modelo que representa una reserva de cancha realizada por un usuario")
public class Reserva {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Schema(description = "Identificador único de la reserva", example = "1")
    private Long id;
    
    @Column(name = "usuario_id", nullable = false)
    @Schema(description = "ID del usuario que realiza la reserva", example = "1", requiredMode = Schema.RequiredMode.REQUIRED)
    private Long usuarioId;
    
    @Column(name = "cancha_id", nullable = false)
    @Schema(description = "ID de la cancha reservada", example = "1", requiredMode = Schema.RequiredMode.REQUIRED)
    private Long canchaId;
    
    @Column(name = "fecha_inicio", nullable = false)
    @Schema(description = "Fecha y hora de inicio de la reserva", example = "2024-01-15T10:00:00", requiredMode = Schema.RequiredMode.REQUIRED)
    private LocalDateTime fechaInicio;
    
    @Column(name = "fecha_fin", nullable = false)
    @Schema(description = "Fecha y hora de fin de la reserva", example = "2024-01-15T12:00:00", requiredMode = Schema.RequiredMode.REQUIRED)
    private LocalDateTime fechaFin;
    
    @Column(nullable = false)
    @Schema(description = "Precio total de la reserva", example = "100000.00", requiredMode = Schema.RequiredMode.REQUIRED)
    private BigDecimal precioTotal;
    
    @Column(length = 50)
    @Enumerated(EnumType.STRING)
    @Schema(description = "Estado actual de la reserva", example = "PENDIENTE", allowableValues = {"PENDIENTE", "CONFIRMADA", "CANCELADA", "COMPLETADA"})
    private EstadoReserva estado;
    
    @Column(length = 500)
    @Schema(description = "Observaciones adicionales sobre la reserva", example = "Reserva para partido de fútbol")
    private String observaciones;
    
    @Column(name = "fecha_creacion")
    @Schema(description = "Fecha y hora de creación de la reserva", example = "2024-01-15T09:00:00")
    private LocalDateTime fechaCreacion;
    
    @Column(name = "fecha_actualizacion")
    @Schema(description = "Fecha y hora de la última actualización", example = "2024-01-15T09:00:00")
    private LocalDateTime fechaActualizacion;
    
    @Schema(description = "Estados posibles de una reserva")
    public enum EstadoReserva {
        @Schema(description = "Reserva creada pero aún no confirmada")
        PENDIENTE,
        @Schema(description = "Reserva confirmada y activa")
        CONFIRMADA,
        @Schema(description = "Reserva cancelada")
        CANCELADA,
        @Schema(description = "Reserva completada (ya se utilizó)")
        COMPLETADA
    }
    
    @PrePersist
    protected void onCreate() {
        fechaCreacion = LocalDateTime.now();
        fechaActualizacion = LocalDateTime.now();
        if (estado == null) {
            estado = EstadoReserva.PENDIENTE;
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        fechaActualizacion = LocalDateTime.now();
    }
}

File: TuCancha_Microservicios/ANDROID_INTEGRATION/models/Reserva.kt
package com.tucancha.data.models

import com.google.gson.annotations.SerializedName

/**
 * Modelo de Reserva para el microservicio Reservas (Puerto 8084)
 */
data class Reserva(
    @SerializedName("id")
    val id: Long? = null,
    
    @SerializedName("usuarioId")
    val usuarioId: Long,
    
    @SerializedName("canchaId")
    val canchaId: Long,
    
    @SerializedName("fechaInicio")
    val fechaInicio: String,  // Formato ISO 8601: "2024-01-15T10:00:00"
    
    @SerializedName("fechaFin")
    val fechaFin: String,
    
    @SerializedName("precioTotal")
    val precioTotal: Double,
    
    @SerializedName("estado")
    val estado: EstadoReserva? = EstadoReserva.PENDIENTE,
    
    @SerializedName("observaciones")
    val observaciones: String? = null,
    
    @SerializedName("fechaCreacion")
    val fechaCreacion: String? = null,
    
    @SerializedName("fechaActualizacion")
    val fechaActualizacion: String? = null
)

enum class EstadoReserva {
    PENDIENTE,
    CONFIRMADA,
    CANCELADA,
    COMPLETADA
}

/**
 * Request para crear reserva (campos mínimos)
 */
data class CrearReservaRequest(
    val usuarioId: Long,
    val canchaId: Long,
    val fechaInicio: String,
    val fechaFin: String,
    val precioTotal: Double,
    val observaciones: String? = null
)

